### PR TITLE
[Arc] Add option to observe registers and memories

### PR DESCRIPTION
**Pull request by fabianschuiki**
Instead of always making registers and memories observable for the user, add dedicated `--observe-registers` and `--observe-memories` options to `arcilator`. These allow the user to indicate whether they are interested in observing register and memory state. If these options are set to false, don't attach "name" or "names" attributes to states and memories, such that other parts of the pipeline have a chance to delete unnecessary state.

### CODE REVIEW

File: include/circt/Conversion/ConvertToArcs.h [hunk 1]
```diff
@@ -13,7 +13,13 @@
 #include <memory>
 
 namespace circt {
-std::unique_ptr<OperationPass<ModuleOp>> createConvertToArcsPass();
+
+#define GEN_PASS_DECL_CONVERTTOARCS
+#include "circt/Conversion/Passes.h.inc"
+
+std::unique_ptr<OperationPass<ModuleOp>>
+createConvertToArcsPass(const ConvertToArcsOptions &options = {});
```
Comment (uenoku) [inline include/circt/Conversion/ConvertToArcs.h:21]: nit: This should be automatically generated if we don't define `constructor` in ODS (Passes.td Line 615)

**Review by TaoBi22 — APPROVED** [review]
LGTM!

File: include/circt/Conversion/ConvertToArcs.h (hunk 1 above)
Comment (fabianschuiki) [reply include/circt/Conversion/ConvertToArcs.h:21]: I did not know that was a thing. That's fantastic! Feels like we can get rid of almost all of those constructors that way 😍

Comment (fabianschuiki) [reply include/circt/Conversion/ConvertToArcs.h:21]: Hmmm, the auto-generated constructor is just called `createConvertToArcs` :crying_cat_face: So we'd have to go through all pass names defined in tablegen files and add a "Pass" suffix.

File: lib/Conversion/ConvertToArcs/ConvertToArcs.cpp [hunk 2]
```diff
@@ -305,7 +311,8 @@ LogicalResult Converter::absorbRegs(HWModuleOp module) {
       // If we absorb a register into the arc, the arc effectively produces that
       // register's value. So if the register had a name, ensure that we assign
       // that name to the arc's output.
-      absorbedNames[result.getResultNumber()] = regOp.getNameAttr();
+      absorbedNames[result.getResultNumber()] =
+          tapRegisters ? regOp.getNameAttr() : emptyNameAttr;
```
[the remark this answers is not in the document]
Comment (fabianschuiki) [reply lib/Conversion/ConvertToArcs/ConvertToArcs.cpp:315]: Excellent point. That would be a lot smarter :innocent:

**Pull request merged**